Alerting: Refactor readiness check

gotjosh opened the pull request:
Moves the readiness check to the mimir client and removes the need to assert that we have senders - it already has a queue and can hold notifications until we're ready to send them.

Signed-off-by: gotjosh <josue.abreu@gmail.com>

santihernandezc commented on pkg/services/ngalert/remote/client/alertmanager.go:
```diff
@@ -0,0 +1,111 @@
+	amclient.AlertmanagerAPI
+}
+
+const alertmanagerAPIMounthPath = "/alertmanager"
```
```suggestion
const alertmanagerAPIMountPath = "/alertmanager"
```

santihernandezc commented on pkg/services/ngalert/remote/client/alertmanager.go:
```diff
@@ -0,0 +1,111 @@
+	amclient "github.com/prometheus/alertmanager/api/v2/client"
+)
+
+type AlertmanagerClient interface {
+	amclient.AlertmanagerAPI
+}
```
Where would this interface be used?

gotjosh replied:
Good shout, I was playing around but couldn't figure out a way to make it work.

santihernandezc commented on pkg/services/ngalert/remote/client/alertmanager.go:
```diff
@@ -0,0 +1,111 @@
+			}
+
+			return true, nil
+		case <-time.After(10 * time.Second):
+			return false, errors.New("ready check timed out")
+		}
```
Remember to cancel any inflight HTTP request here.

gotjosh committed:
address review comments

Signed-off-by: gotjosh <josue.abreu@gmail.com>

gotjosh committed:
Alerting: Refactor readiness check

Moves the readiness check to the mimir client and removes the need to assert that we have senders - it already has a queue and can hold notifications until we're ready to send them.

Signed-off-by: gotjosh <josue.abreu@gmail.com>

gotjosh committed:
dcoument the function

Signed-off-by: gotjosh <josue.abreu@gmail.com>

gotjosh committed:
Fix tests

Signed-off-by: gotjosh <josue.abreu@gmail.com>

gotjosh committed:
make the am ready

Signed-off-by: gotjosh <josue.abreu@gmail.com>

santihernandezc commented on pkg/services/ngalert/remote/client/alertmanager.go:
```diff
@@ -0,0 +1,113 @@
+			attempts++
+			status, err := attempt()
+			if err != nil {
+				am.logger.Error("Ready check attempt failed", "attempt", attempts, "err", err)
```
I don't think we should log each failed attempt. If we fail for the entire 10 seconds we'll flood the output with <=100 error level log lines.

gotjosh replied:
I can change them to debug if you prefer, but we certainly don't expect this to happen very often and I think we'd like to know when it does to make sure we can correct it.

The alertmanager is designed with availability in mind - so even under unstable network conditions, we don't expect constant failures unless there's an outage.

santihernandezc commented on pkg/services/ngalert/remote/client/alertmanager.go:
```diff
@@ -0,0 +1,113 @@
+			}
+
+			return true, nil
+		case <-time.After(10 * time.Second):
```
`time.After()` is creating being evaluated and creating a new channel on each loop iteration, we're retrying infinitely. We should create the timer outside of this loop:
```go
// Outside.
timeout := time.After(10 * time.Second)

// Inside
case <-timeout:
	cancel()
```


Since we're using a context for the readiness check, the easiest option would be to use a `context.WithTimeout(10 * time.Second)` instead of `context.WithCancel()`. We'd skip creating the timer and canceling the context on timeout.

gotjosh replied:
> time.After() is creating being evaluated and creating a new channel on each loop iteration, we're retrying infinitely. 

Thanks, good spot.

> Since we're using a context for the readiness check, the easiest option would be to use a context.WithTimeout(10 * time.Second) instead of context.WithCancel(). 

Semantically, I don't think this is correct even if they both achieve the same. `withTimeout` is typically used in the operation you're using the `ctx` on is meant to have a deadline. However, in our case the atomic operation (executing the HTTP request) is less important than the combination of them.

In essence, is different to say - I care about this HTTP call taking 10s vs I care about exceeding 10s while I make repeated HTTP calls. 

`WithTimeout` is commonly used on the former - so I'll stick with the current approach.

santihernandezc replied:
Ah, nice to know! Go for the other approach then.

santihernandezc commented on pkg/services/ngalert/remote/client/alertmanager.go:
```diff
@@ -0,0 +1,113 @@
+// IsReadyWithBackoff executes a readiness check against the `/-/ready` Alertmanager endpoint.
+// If it takes more than 10s to get a response back - we abort the check.
+func (am *Alertmanager) IsReadyWithBackoff(ctx context.Context) (bool, error) {
+	ctx, cancel := context.WithCancel(ctx)
```
We can skip creating the timer below.
```suggestion
	ctx, cancel := context.WithTimeout(ctx, 10 * time.Second)
```

gotjosh replied:
Ditto from above.

santihernandezc commented on pkg/services/ngalert/remote/client/alertmanager.go:
```diff
@@ -0,0 +1,113 @@
+			}
+
+			return true, nil
+		case <-time.After(10 * time.Second):
+			cancel()
+			return false, errors.New("ready check timed out")
```
```suggestion
		case <-ctx.Done():
			return false, errors.New("ready check timed out")
```

If we decide to not emit a log line for each failed attempt, we could save the last error in a variable and wrap it here.

gotjosh committed:
address review feedback

Signed-off-by: gotjosh <josue.abreu@gmail.com>

gotjosh committed:
fix url

Signed-off-by: gotjosh <josue.abreu@gmail.com>

gotjosh committed:
In tests, we need to make sure we're ready to receive things.

Signed-off-by: gotjosh <josue.abreu@gmail.com>

santihernandezc reviewed (approved):
LGTM 👍
